bson/src/main/org/bson/BsonBinary.java

```diff
@@ -89,6 +92,20 @@ public BsonBinary(final UUID uuid) {
         this(uuid, UuidRepresentation.STANDARD);
     }
 
+    /**
```

Comment:
I'm wondering if we should add similar support for vectors to `org.bson.types.Binary`, since that's the class more typically used with `org.bson.Document`.

I know we didn't do that for UUIDs, but can't remember the rationale.

What do you think?

Reply:
Yes, we should, as it would otherwise limit users who work with the `org.bson.Document` class. I've added the support for vectors in `org.bson.types.Binary.` Thank you!

Reply:
Good questions, @vbabanin.  

> Why do those types exist

These types go way back to 2009, five year before we decided to create the BsonValue type hierarchy.  Back then, the way to create BSON documents was with a class called BasicDBObject (the precursor to Document), which is a `Map<String, Object>`.  Like Document, you could add Integer/Long/Date, but then we created classes like ObjectId, Binary, etc. for the types that didn't have a good analog in the JDK.

The types in the BsonValue hierarchy, on the other hand, were meant to be used with BsonDocument, though in practice they can be used with Document as well.  However, in places like https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/data-formats/documents/ we list Binary and not BsonBinary as what an application would typically use with Document.

> Not only it does not add anything to the functionality of org.bson.BsonBinary, but it does not support UUIDs similarly to how BsonBinary does. Given this, why does it need to support Vector?

I think the reason we didn't add UUID support to Binary is that you can just put a UUID directly into a Document, whereas you can't do that with a BsonDocument.

Reply:
I now realize that we could do the same thing for Vector as we do for UUID: allow it to be serialized from and deserialized into a Document.  If we do that, then we could remove the Vector support from Binary.  But that won't just work out of the box.  The issue is round-tripping.  You can add a Vector to a Document and it will be encoded properly, but what happens during decoding? The `BsonTypeClassMap#DEFAULT_BSON_TYPE_CLASS_MAP` used by `DocumentCodec` will decode it as `Binary`, not Vector.  To make round-tripping work for UUIDs, there is an ugly switch statement in `ContainerCodecHelper#readValue` which would have to be adjusted to account for Vector.

So we have to decide whether that's the behavior that we want for Vector, or whether we should require users to manually convert between Binary and Vector.  UUID has worked this way since the dawn of time, but we can make a different decision for Vector if we want.

Reply:
We talked about that exact idea with @vbabanin yesterday, and we decided to have the same behavior of `Vector` as it is for `UUID`. The issue is that round-tripping is broken either way:

- option one: encode `BsonBinary`/`Binary`/`Vector`, but decode `Vector`;
- option two: encode `BsonBinary`/`Binary`/`Vector`, but decode `BsonBinary`/`Binary`.

It looks like there are always situations when we cannot decode the Java type that was used when encoding.

Reply:
Yes, there is not enough information encoded in BSON to round trip.  We have to choose a canonical type to decode into regardless of what was used to encode.

> option two: encode BsonBinary/Binary/Vector, but decode BsonBinary/Binary.

Given that all other BSON binary are decoded as `Binary`, I think it would be unexpected to decode vectors as `BsonBinary`.

Reply:
OK, looks like you with the latest patch, any representation of Vector can be encoded b DocumentCodec: Vector subclasses, Binary, BsonBinary.  But vectors are always decoded to one of the Vector subclasses.

sgtm, resolving.